Feat/backend #8 job description generator

Alvitasy opened the pull request:
### Pull Request: Backend – Job Description Generator

Addresses #8 

---

### What This PR Does

This PR implements the backend support for a **Job Description Generator** using FastAPI and Gemini LLM.  
It provides an API endpoint to generate job descriptions from structured inputs.

---

###  Why This Is Useful

The endpoint helps streamline the job description creation process by auto-generating professional, structured descriptions for recruiters or hiring managers.  
It includes built-in validation and formatting rules to ensure clarity, professionalism, and consistency.

---

### Summary of Changes

- **`router.py`**: Added a new POST endpoint `/job-descriptions/` to handle job description generation requests via FastAPI.
- **`schemas.py`**: Defined `JobDescriptionRequest` and `JobDescriptionResponse` models using Pydantic for structured request/response handling.
- **`service.py`**: Implemented the `generate_job_description()` function which:
  - Validates input using `validator.py`
  - Constructs prompts with `prompts.py`
  - Calls Gemini LLM through `generate_text()`
- **`prompts.py`**: Created markdown-based prompt templates (`JOB_DESCRIPTION_TEMPLATE`, `JOB_BENEFITS_SECTION`, and `PROMPT_SUFFIX`) with formatting rules and content guardrails.
- **`exceptions.py`**: Introduced custom exceptions `InvalidJobInputError` and `JobDescriptionServiceError` for clean API error handling.
- **`validator.py`**: Built `is_irrelevant()` function to detect and reject low-quality or placeholder inputs using regex and known junk tokens.



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Enhanced job description generation with support for custom notes, key focus areas, and benefits.
  - Improved input validation with clear error messages for invalid requirements.

- **Bug Fixes**
  - More robust error handling and structured logging for job description service exceptions.

- **Refactor**
  - Updated request and response formats for job descriptions, including more descriptive field explanations.
  - Replaced and relocated the job description prompt template for better maintainability.

- **Chores**
  - Removed obsolete code and unused imports to streamline the codebase.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Alvitasy committed:
feat(frontend): add Job Description Generator UI logic #8

Alvitasy committed:
feat(backend): implement Job Description Generator #8

Alvitasy committed:
revert: frontend Job_Description.py changes from backend branch

This reverts commit 319ae3da420e9f1d38e255074af30ce8fdb04ff7, which added frontend changes to the backend feature branch by mistake.

These changes have been moved to the correct frontend branch instead.

sanjeevp-7t commented on backend/job_descriptions/prompts.py:
```diff
@@ -1,4 +1,53 @@
+- Follow the section structure exactly as provided.
+- Use bullet points for clarity, and avoid repetition.
+- Do not include salary information unless explicitly specified.
+- If the {custom_note} includes terms such as “undergrad student”, “opt student”, “cpt student”, or “f1 student” (not case-sensitive), include relevant notes.
```
You probably don't want to include the "custom_note" input here. Rather, just saying if a custom note is included, do XYZ.


```suggestion
- If the "custom_note" includes terms such as “undergrad student”, “opt student”, “cpt student”, or “f1 student” (not case-sensitive), include relevant notes.
```

sanjeevp-7t commented on backend/job_descriptions/prompts.py:
```diff
@@ -1,4 +1,53 @@
+   - List 10–12 specific technical and soft skills related to **{key_focus}**.
+   - Include tools, technologies, methodologies, and relevant interpersonal skills.
+   - Each bullet point should explain why the skill is important or how it's applied in the role.
+""".strip()  # noqa: RUF001
```
The current code uses Unicode em dashes (—) which can cause encoding issues and make the code less portable. We should use regular ASCII hyphens (-) instead.

Please replace all Unicode dashes with regular ASCII hyphens (-) and remove the RUF001 ignore comment.


```suggestion
1. **About the Job**
   - Start with a 2-3 sentence overview of the role, including the team or department context.
   - Then include this custom note: "{custom_note}"
   - Highlight key responsibilities, tools, or impact areas in 3-5 bullet points.
   - Each bullet point should be 1-2 sentences long to provide clear, descriptive context.
   - Use professional language with specifics where applicable.
   - Avoid generic phrases like "work on exciting projects" or "be part of a dynamic team." Instead, focus on tangible responsibilities and technologies used.
   - If experience level is specified, tailor the description to that level (e.g., 1 yr, 7 yr).

2. **Required Skills**
   - List 10-12 specific technical and soft skills related to **{key_focus}**.
   - Include tools, technologies, methodologies, and relevant interpersonal skills.
   - Each bullet point should explain why the skill is important or how it's applied in the role.
""".strip()
```

sanjeevp-7t commented on backend/job_descriptions/prompts.py:
```diff
@@ -1,4 +1,53 @@
+""".strip()
+
+# OpenAI-style simple prompt (e.g., used in fallback models)
+OPENAI_JOB_DESCRIPTION_PROMPT = (
```
Remove this prompt, if it is not being used.

sanjeevp-7t commented on backend/job_descriptions/router.py:
```diff
@@ -13,15 +16,22 @@
 
 
-@router.post("/")
+@router.post("/", response_model=JobDescriptionResponse)  # noqa: FAST001
```
The "response_model" parameter is redundant when the return type is already annotated. FastAPI infers it automatically. Remove the parameter and the # noqa: FAST001 comment to let the linter catch these in the future.


```suggestion
@router.post("/")
```

sanjeevp-7t commented on backend/job_descriptions/exceptions.py:
```diff
@@ -11,3 +11,15 @@ def __init__(
+        status_code: int = 400,
+    ) -> None:
+        """Initialize the InvalidJobInputError with a detail message and status code."""
+        super().__init__(status_code=status_code, detail=detail)
```
Helps to be specific when naming exception classes.

```suggestion
class InvalidJobRequirementsError(HTTPException):
    """Raised when the input to the job description service is irrelevant or inappropriate."""

    def __init__(
        self,
        detail: str = "The job title or provided information does not appear valid for generating a job description.",
        status_code: int = 400,
    ) -> None:
        """Initialize the InvalidJobRequirementsError with a detail message and status code."""
        super().__init__(status_code=status_code, detail=detail)
```

sanjeevp-7t commented on backend/job_descriptions/router.py:
```diff
@@ -1,6 +1,9 @@
 
-from backend.job_descriptions.exceptions import JobDescriptionServiceError
+from backend.job_descriptions.exceptions import (
+    InvalidJobInputError,
```
```suggestion
    InvalidJobRequirementsError,
```

sanjeevp-7t commented on backend/job_descriptions/router.py:
```diff
@@ -13,15 +16,22 @@
-        return JobDescriptionResponse(job_description=generated_description)
+        return JobDescriptionResponse(job_description=description)
+
+    except InvalidJobInputError:
```
```suggestion
    except InvalidJobRequirementsError:
```

sanjeevp-7t commented on backend/job_descriptions/schemas.py:
```diff
@@ -2,19 +2,33 @@
 
 class JobDescriptionRequest(BaseModel):
-    """Request model for generating a job description."""
+    """Input model for generating a job description."""
```
```suggestion
    """Request model for generating a job description."""
```

sanjeevp-7t commented on backend/job_descriptions/schemas.py:
```diff
@@ -2,19 +2,33 @@
 
 class JobDescriptionResponse(BaseModel):
-    """Response model for a generated job description."""
+    """Output model containing the generated job description."""
```
```suggestion
    """Response model containing the generated job description."""
```

sanjeevp-7t commented on backend/job_descriptions/router.py:
```diff
@@ -13,15 +16,22 @@
+            job_title=request.title,
+            custom_note=request.custom_note,
+            key_focus=request.key_focus,
+            benefits=request.benefits or "",
```
`request.benefits` already defaults to None, if it is not provided in the request.

```suggestion
            benefits=request.benefits
```

sanjeevp-7t commented on backend/job_descriptions/service.py:
```diff
@@ -1,8 +1,42 @@
+
+
+async def generate_job_description(
+    job_title: str, custom_note: str, key_focus: str, benefits: str = ""
```
```suggestion
    job_title: str, custom_note: str, key_focus: str, benefits: str
```

sanjeevp-7t commented on backend/job_descriptions/validator.py:
Hardcoding validation rules against the job description input is not scalable nor efficient. It's better to add guardrails like these in your prompt, instead.

sanjeevp-7t commented on backend/job_descriptions/service.py:
```diff
@@ -1,8 +1,42 @@
+) -> str:
+    """Generate a structured, markdown-formatted job description using an LLM."""
+    # Input validation for professionalism and relevance
+    if any(map(is_irrelevant, [job_title, custom_note, key_focus])):
+        raise InvalidJobInputError
```
Look at my comment in `validator.py`.

sanjeevp-7t commented on backend/job_descriptions/prompts.py:
```diff
@@ -1,4 +1,53 @@
-JOB_DESCRIPTION_PROMPT = (
+# This module contains the prompt templates for generating job descriptions.
+# Main job description template
+JOB_DESCRIPTION_TEMPLATE = """
```
This is more of a prompt than a template.

```suggestion
JOB_DESCRIPTION_PROMPT = """
```

sanjeevp-7t commented on backend/job_descriptions/prompts.py:
- Rename this file to `constants.py`.
- Might be better to ask the LLM to return a JSON output, if you are looking to identify between success and error responses. For example:
```
{ job_description: "markdown-formatted job description" }
{ error: "error message" }
```

sanjeevp-7t commented on backend/job_descriptions/service.py:
```diff
@@ -1,8 +1,42 @@
+        )
+
+        # Optionally add benefits section
+        if benefits.strip():
```
```suggestion
        if benefits:
```

Alvitasy committed:
feat(frontend/job-description): add JD generator UI in backend branch

Alvitasy committed:
Revert "feat(frontend/job-description): add JD generator UI in backend branch"

This reverts commit c304a6119a7565e911550f78d79fd174c731a92b.

Alvitasy committed:
feat(backend): enhance job description generation logic

- Updated job description prompt to enforce strict JSON format and improved input validation
- Simplified error handling in the job description creation endpoint
- Removed the validator module and integrated it directly into the prompt
- Enhanced the extraction of JSON from LLM responses to ensure proper formatting and error management

Files changed:
- backend/job_descriptions/service.py
- backend/job_descriptions/schemas.py
- backend/job_descriptions/router.py
- backend/job_descriptions/exceptions.py

Addresses #8

Alvitasy committed:
feat(backend): address review feedback and apply stashed changes

- Renamed `prompts.py` to `constants.py` for clarity and maintainability.
- Renamed `InvalidJobInputError` to `InvalidJobRequirementsError` for clarity in error handling.
- Adjusted docstrings in request and response models for consistency.

Files changed:
- backend/job_descriptions/constants.py
- backend/job_descriptions/exceptions.py
- backend/job_descriptions/schemas.py
- backend/job_descriptions/service.py

Alvitasy committed:
feat(backend): clean up prompts.py after renaming to constants.py

- Removed prompts.py after it was renamed to constants.py during the job description module refactor.
- Ensures codebase consistency and avoids redundancy following recent improvements.

Files changed:
- backend/job_descriptions/prompts.py

Addresses #8
